cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java

```diff
@@ -222,6 +229,30 @@ public void start() {
     leaderElection.startAndWait();
   }
 
+  /**
+   * programs jars are copied to twill directory, for secure impersonation to work,
+   * we want other users to be able to write to the twill directory,
+   * currently only cdap.user has read-write permissions while other users can only read the "/cdap/twill" dir,
+   * we want to let others write to twill directory, till we have a better solution.
+   */
+  private void createTwillDirectory() {
+    try {
+      String twillPath = String.format("/%s/twill", cConf.get(Constants.ROOT_NAMESPACE));
+      FsPermission permission = new FsPermission(FsAction.ALL, FsAction.ALL, FsAction.ALL);
+      FileContext fileContext = FileContext.getFileContext(hConf);
+      // file context does ( permission AND  (NOT of umask) ) and uses that as permission, by default umask is 022,
+      // if we want 777 permission, we have to set umask to 000
+      fileContext.setUMask(new FsPermission(FsAction.NONE, FsAction.NONE, FsAction.NONE));
+      fileContext.mkdir(new org.apache.hadoop.fs.Path(twillPath), permission, false);
+    } catch (AccessControlException | ParentNotDirectoryException |
+      FileNotFoundException | FileAlreadyExistsException e) {
+      // just log the exception
+      LOG.error("Exception while trying to twill directory {}", e);
```

Comment:
if file already exists, we can check the permission configured for other users for that file, if its not  writable for others, we can log an error message that "secure impersonation might not work if its not writable for others, but still not throw an exception".